File: frontend/src/api/rooms.ts
import { apiFetch } from './client';
import { DecisionRoom, DecisionRoomWithStages, RoomStage } from '../types';

export function createRoom(workspaceId: string, title: string, description?: string) {
  return apiFetch<DecisionRoomWithStages>(`/api/workspaces/${workspaceId}/rooms`, {
    method: 'POST',
    body: JSON.stringify({ title, description }),
  });
}

export function listRooms(workspaceId: string) {
  return apiFetch<DecisionRoom[]>(`/api/workspaces/${workspaceId}/rooms`);
}

export function getRoom(roomId: string) {
  return apiFetch<DecisionRoomWithStages>(`/api/rooms/${roomId}`);
}

export function updateRoom(roomId: string, data: { title?: string; description?: string }) {
  return apiFetch<DecisionRoom>(`/api/rooms/${roomId}`, {
    method: 'PATCH',
    body: JSON.stringify(data),
  });
}

export function transitionStage(roomId: string) {
  return apiFetch<DecisionRoomWithStages>(`/api/rooms/${roomId}/transition`, {
    method: 'POST',
  });
}

export function decide(roomId: string, proposalId: string, summary?: string) {
  return apiFetch<DecisionRoom>(`/api/rooms/${roomId}/decide`, {
    method: 'POST',
    body: JSON.stringify({ proposal_id: proposalId, summary }),
  });
}

export function getStages(roomId: string) {
  return apiFetch<RoomStage[]>(`/api/rooms/${roomId}/stages`);
}

export function replaceStages(roomId: string, stages: { name: string; stage_type: string }[]) {
  return apiFetch<RoomStage[]>(`/api/rooms/${roomId}/stages`, {
    method: 'PUT',
    body: JSON.stringify(stages),
  });
}

File: backend/src/handlers/rooms.rs
use axum::extract::{Path, State};
use axum::Json;
use uuid::Uuid;

use crate::auth::middleware::AuthUser;
use crate::error::AppError;
use crate::models::room::{CreateRoomInput, DecideInput, DecisionRoom, DecisionRoomWithStages, UpdateRoomInput};
use crate::models::stage::{RoomStage, StageInput};
use crate::services::{room_service, stage_service};
use crate::AppState;

pub async fn create_room(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(workspace_id): Path<Uuid>,
    Json(input): Json<CreateRoomInput>,
) -> Result<Json<DecisionRoomWithStages>, AppError> {
    let room = room_service::create_room(&state.db, workspace_id, auth.user_id, input).await?;
    Ok(Json(room))
}

pub async fn list_rooms(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(workspace_id): Path<Uuid>,
) -> Result<Json<Vec<DecisionRoom>>, AppError> {
    let rooms = room_service::list_rooms(&state.db, workspace_id, auth.user_id).await?;
    Ok(Json(rooms))
}

pub async fn get_room(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(room_id): Path<Uuid>,
) -> Result<Json<DecisionRoomWithStages>, AppError> {
    let room = room_service::get_room(&state.db, room_id, auth.user_id).await?;
    Ok(Json(room))
}

pub async fn update_room(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(room_id): Path<Uuid>,
    Json(input): Json<UpdateRoomInput>,
) -> Result<Json<DecisionRoom>, AppError> {
    let room = room_service::update_room(&state.db, room_id, auth.user_id, input).await?;
    Ok(Json(room))
}

pub async fn transition_stage(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(room_id): Path<Uuid>,
) -> Result<Json<DecisionRoomWithStages>, AppError> {
    let room = room_service::transition_stage(&state.db, room_id, auth.user_id).await?;
    Ok(Json(room))
}

pub async fn decide(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(room_id): Path<Uuid>,
    Json(input): Json<DecideInput>,
) -> Result<Json<DecisionRoom>, AppError> {
    let room = room_service::decide(&state.db, room_id, auth.user_id, input).await?;
    Ok(Json(room))
}

pub async fn get_stages(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(room_id): Path<Uuid>,
) -> Result<Json<Vec<RoomStage>>, AppError> {
    let stages = stage_service::get_stages(&state.db, room_id, auth.user_id).await?;
    Ok(Json(stages))
}

pub async fn replace_stages(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(room_id): Path<Uuid>,
    Json(stages): Json<Vec<StageInput>>,
) -> Result<Json<Vec<RoomStage>>, AppError> {
    let stages = stage_service::replace_stages(&state.db, room_id, auth.user_id, stages).await?;
    Ok(Json(stages))
}
